### DISCUSSION TITLE
mac version 11.6 , M1,  brew install openjdk@11

### DISCUSSION THREAD

**Discussion by antyang26** [Homebrew/discussions#2405] [Everyday usage]
1. brew install openjdk@11 
2. openjdk@11 11.0.12 is already installed 
3. java -version
 error:
 The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.

???

**Comment by antyang26** [comment]
![image](https://user-images.githubusercontent.com/28669095/140489030-301dc8bc-dd0f-4db7-9ba5-db2ddf629a73.png)
![image](https://user-images.githubusercontent.com/28669095/140489042-a11a7b3b-db8a-4832-9b3c-144caf28b42e.png)

**Comment by SMillerDev** [comment]
The install warns you about this:

```
For the system Java wrappers to find this JDK, symlink it with
  sudo ln -sfn /opt/homebrew/opt/openjdk@11/libexec/openjdk.jdk /Library/Java/JavaVirtualMachines/openjdk-11.jdk

openjdk@11 is keg-only, which means it was not symlinked into /opt/homebrew,
because this is an alternate version of another formula.
```

You'll need to call this as `/opt/homebrew/opt/openjdk@11/bin/java`

**Reply by antyang26** [reply]
thanks,

**Reply by ni3stack** [reply]
Hi @antyang26 @SMillerDev . I'm new to the MAC os environment.

my Mac os- Monetery 12.1 , M1.

I used brew to install maven, which also installed openjdk18.
while mvn -v works fine.. java -version gives the same error as shown in the above-attached screenshots.

Install warns me about symlink.. but I don't get the context.
Could you please help me understand it?

which java prints /usr/bin/java and
when i tried doing /usr/libexec/java_home got below error


The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.
install warns mentioned JavaVirtualMachines -but there is no openjdk-XX.jdk in /Library/Java/JavaVirtualMachines

/Library/Java/JavaVirtualMachines -> this folder is empty in my system.

homebrew installed mvn and openjdk in

opt/homebrew/cellar/mvn
opt/homebrew/cellar/openjdk.

what do I have to set to make java work?

**Reply by SMillerDev** [reply]
Run `brew info openjdk` and follow the instructions there.

**Reply by ni3stack** [reply]
@SMillerDev 
do I have to perform all the highlighted steps (marked bold)?

openjdk: stable 18 (bottled) [keg-only]
Development kit for the Java programming language
https://openjdk.java.net/
/opt/homebrew/Cellar/openjdk/18 (642 files, 308.9MB)
  Poured from bottle on 2022-03-30 at 13:54:06
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/openjdk.rb
License: GPL-2.0-only with Classpath-exception-2.0

==> Dependencies
Build: autoconf ✘
==> Caveats
For the system Java wrappers to find this JDK, symlink it with
  **sudo ln -sfn /opt/homebrew/opt/openjdk/libexec/openjdk.jdk /Library/Java/JavaVirtualMachines/openjdk.jdk**

openjdk is keg-only, which means it was not symlinked into /opt/homebrew,
because macOS provides similar software and installing this software in
parallel can cause all kinds of trouble.

If you need to have openjdk first in your PATH, run:
  **echo 'export PATH="/opt/homebrew/opt/openjdk/bin:$PATH"' >> /Users/nitpatha/.bash_profile**

For compilers to find openjdk you may need to set:
  **export CPPFLAGS="-I/opt/homebrew/opt/openjdk/include"**

**Reply by SMillerDev** [reply]
> For the system Java wrappers to find this JDK, symlink it with
> sudo ln -sfn /opt/homebrew/opt/openjdk/libexec/openjdk.jdk /Library/Java/JavaVirtualMachines/openjdk.jdk

This is the system wrapper you were talking about here:
>  when i tried doing /usr/libexec/java_home got below error

If you need to have OpenJDK first in your path, you can run the instruction for that, and if you're compiling some C-family program against java you can run the last command you highlighted.

**Reply by ni3stack** [reply]
ok, so I ran the path option setup, and it's working... like I can see the `java -v` details.

I'm not compiling any C-family program, it's for a maven project.

correct me if I'm wrong,  in order to make `/usr/libexec/java_home` work we need to go with the symlink option?

system java wrappers by default look for jdk in ` /Library/Java/JavaVirtualMachines/` and since nothing is there so it will check in 
`howbrew/opt/openjdk/libexec/openjdk.jdk` as it's been symlinked?

**Reply by SMillerDev** [reply]
Yes, that is correct.

**Reply by ni3stack** [reply]
Thank you!

**Comment by wweevv-johndpope** [comment]
For sanity check - I had 18 installed - but couldn't toggle down to 11.
check your JAVA_HOME - https://mkyong.com/java/how-to-set-java_home-environment-variable-on-mac-os-x/



```shell
/usr/libexec/java_home -V
Matching Java Virtual Machines (3):
    18.0.2 (arm64) "Homebrew" - "OpenJDK 18.0.2" /opt/homebrew/Cellar/openjdk/18.0.2/libexec/openjdk.jdk/Contents/Home
    18.0.1 (arm64) "Eclipse Adoptium" - "OpenJDK 18.0.1" /Library/Java/JavaVirtualMachines/temurin-18.jdk/Contents/Home
    11.0.16 (arm64) "Homebrew" - "OpenJDK 11.0.16" /opt/homebrew/Cellar/openjdk@11/11.0.16/libexec/openjdk.jdk/Contents/Home
/opt/homebrew/Cellar/openjdk/18.0.2/libexec/openjdk.jdk/Contents/Home
```
~/.zshenv ADD THIS 
```
export JAVA_HOME=$(/usr/libexec/java_home -v11)
```

```
echo $JAVA_HOME
/opt/homebrew/Cellar/openjdk@11/11.0.16/libexec/openjdk.jdk/Contents/Home
java --version
openjdk 11.0.16 2022-07-19
OpenJDK Runtime Environment Homebrew (build 11.0.16+0)
OpenJDK 64-Bit Server VM Homebrew (build 11.0.16+0, mixed mode)
```

**Reply by serey168** [reply]
🙏 Thanks !

**Reply by mertse7en** [reply]
thanks
